### PR TITLE
Enhancement/records import

### PR DESCRIPTION
**Pull request by SteRiccio**
solves #1906

### CODE REVIEW

File: core/record/_recordValidator/attributeValidator.js [hunk 1]
```diff
@@ -18,10 +18,13 @@ import * as AttributeTypeValidator from './attributeTypeValidator'
...
-  (NodeDef.isKey(nodeDef) || NodeDefValidations.isRequired(NodeDef.getValidations(nodeDef))) && Node.isValueBlank(node)
-    ? { key: Validation.messageKeys.record.valueRequired }
-    : null
+const _validateRequired =
+  ({ nodeDef }) =>
+  (_propName, node) =>
+    (NodeDef.isKey(nodeDef) || NodeDefValidations.isRequired(NodeDef.getValidations(nodeDef))) &&
+    Node.isValueBlank(node)
+      ? { key: Validation.messageKeys.record.valueRequired }
+      : null
```
Comment (SteRiccio) [inline core/record/_recordValidator/attributeValidator.js:27]: this is only formatting...

File: server/modules/record/repository/nodeRepository.js [hunk 2]
```diff
@@ -172,29 +183,39 @@ export const updateNode = async (surveyId, nodeUuid, value, meta, draft, client
...
-    [JSON.stringify(value), meta || {}, nodeUuid]
-  )
-  const node = await client.one(
-    `
-    ${_getNodeSelectQuery({ surveyId, draft })}
-    WHERE n.uuid = $1
-  `,
-    nodeUuid,
-    dbTransformCallback
+    [A.isEmpty(value) ? null : JSON.stringify(value), meta || {}, nodeUuid]
```
Comment (SteRiccio) [inline server/modules/record/repository/nodeRepository.js:186]: I don't know if it's better to use 'default' or null in that case...

File: server/modules/record/repository/nodeRepository.js [hunk 3]
```diff
@@ -172,29 +183,39 @@ export const updateNode = async (surveyId, nodeUuid, value, meta, draft, client
...
-    WHERE uuid = $1
-    RETURNING *`,
-    [parentNodeUuid],
-    dbTransformCallback
+export const updateNodes = async ({ surveyId, nodes }, client = db) => {
+  const values = nodes.map((node) => {
+    const value = Node.getValue(node)
+    return [
+      Node.getId(node),
+      value === null ? null : JSON.stringify(value),
```
Comment (SteRiccio) [inline server/modules/record/repository/nodeRepository.js:200]: same as above

File: core/record/_record/recordUpdater.js [hunk 4]
```diff
@@ -41,34 +41,32 @@ export const mergeNodes = nodes => record => {
...
-  }
+  recordUpdated[keys.nodes] = { ...RecordReader.getNodes(record) }
 
-  R.forEachObjIndexed((node, nodeUuid) => {
+  Object.entries(nodes).forEach(([nodeUuid, node]) => {
     if (Node.isDeleted(node)) {
       recordUpdated = deleteNode(node)(recordUpdated)
     } else {
       recordUpdated[keys.nodes][nodeUuid] = node
       recordUpdated = NodesIndex.addNode(node)(recordUpdated)
```
Comment (ramirobg94) [inline core/record/_record/recordUpdater.js:53]: what happens if the node previously exist?

File: server/modules/record/manager/_recordManager/nodeUpdateDependentManager.js [hunk 5]
```diff
@@ -27,32 +25,40 @@ const logExpressionError = ({ error, expressionType, survey, nodeDef, expression
  * Module responsible for updating applicable and default values.
  */
 
-export const updateSelfAndDependentsApplicable = async (survey, record, node, tx) => {
+export const updateSelfAndDependentsApplicable = ({ survey, record, node }) => {
```
Comment (ramirobg94) [inline server/modules/record/manager/_recordManager/nodeUpdateDependentManager.js:28]: do you need to pass the transaction?

Comment (SteRiccio) [reply server/modules/record/manager/_recordManager/nodeUpdateDependentManager.js:28]: no! the functions in this module don't need the transaction anymore: they update the passed record in memory. They could be moved to inner functions of the RecordUpdater module.

File: core/record/_record/recordUpdater.js (hunk 4 above)
Comment (SteRiccio) [reply core/record/_record/recordUpdater.js:53]: it's replaced by the passed one

File: server/modules/record/manager/recordManager.js [hunk 6]
```diff
@@ -28,21 +27,21 @@ export const insertRecord = async (user, surveyId, record, system = false, clien
...
+    Node.getRecordUuid(node),
+    Node.getParentUuid(node),
+    Node.getNodeDefUuid(node),
+    JSON.stringify(Node.getValue(node, null)),
+    Node.getMeta(node),
+  ])
+  const activities = nodes.map((node) => ActivityLog.newActivity(ActivityLog.type.nodeCreate, node, true))
+
+  await NodeRepository.insertNodesFromValues(surveyId, nodeValues, tx)
+  await ActivityLogRepository.insertMany(user, surveyId, activities, tx)
```
Comment (ramirobg94) [inline server/modules/record/manager/recordManager.js:44]: is responsibility of this manager?

File: server/modules/record/repository/nodeRepository.js (hunk 2 above)
Comment (ramirobg94) [reply server/modules/record/repository/nodeRepository.js:186]: lets see

File: server/modules/record/manager/recordManager.js (hunk 6 above)
Comment (SteRiccio) [reply server/modules/record/manager/recordManager.js:44]: uhm, usually we use the repositories and the activity log repository in the manager modules...

**Review by ramirobg94 — APPROVED** [review]

**Pull request merged**